File: python/product.py
# product.py

class Product:
    def __init__(self, name, price, stock):
        self.name = name
        self.price = price
        self.stock = stock

    def get_name(self):
        return self.name

    def set_name(self, name):
        self.name = name

    def get_price(self):
        return self.price

    def set_price(self, price):
        self.price = price

    def get_stock(self):
        return self.stock

    def set_stock(self, stock):
        self.stock = stock

    def update_stock(self, quantity):
        if quantity > self.stock:
            print("Not enough in stock..")
        else:
            self.stock -= quantity

    def display(self):
        print(f"Product: {self.name}\nPrice: ${self.price}\nStock: {self.stock}\n")


class Volleyball(Product):
    def __init__(self, name, price, stock, material, size):
        super().__init__(name, price, stock)
        self.material = material
        self.size = size

    def display(self):
        super().display()
        print(f"Material: {self.material}\nSize: {self.size} inches\n")


class VolleyballNet(Product):
    def __init__(self, name, price, stock, net_type):
        super().__init__(name, price, stock)
        self.type = net_type

    def display(self):
        super().display()
        print(f"Type: {self.type}\n")


class Paddle(Product):
    def __init__(self, name, price, stock, style):
        super().__init__(name, price, stock)
        self.style = style

    def display(self):
        super().display()
        print(f"Style: {self.style}\n")


class Table(Product):
    def __init__(self, name, price, stock, material):
        super().__init__(name, price, stock)
        self.material = material

    def display(self):
        super().display()
        print(f"Material: {self.material}\n")


class Ball(Product):
    pass  # Inherits everything from Product


class Tent(Product):
    def __init__(self, name, price, stock, size, durability, weather_rating):
        super().__init__(name, price, stock)
        self.size = size
        self.durability = durability
        self.weather_rating = weather_rating

    def display(self):
        super().display()
        print(f"Size: {self.size}\nDurability: {self.durability}\nWeather Rating: {self.weather_rating}\n")


class SleepingBag(Product):
    def __init__(self, name, price, stock, size):
        super().__init__(name, price, stock)
        self.size = size

    def display(self):
        super().display()
        print(f"Size: {self.size}\n")


class SoccerBall(Product):
    def __init__(self, name, price, stock, material):
        super().__init__(name, price, stock)
        self.material = material

    def display(self):
        super().display()
        print(f"Material: {self.material}\n")


class GoalNet(Product):
    def __init__(self, name, price, stock, size):
        super().__init__(name, price, stock)
        self.size = size

    def display(self):
        super().display()
        print(f"Size: {self.size}\n")


def display_product_details(product):
    if product is None:
        print("Invalid Product.")
        return
    product.display()


def handle_purchase(product, quantity):
    if product.get_stock() >= quantity:
        total_price = product.get_price() * quantity
        print(f"\nTotal Price: ${total_price}")
        product.update_stock(quantity)
        print(f"Updated Stock: {product.get_stock()} remaining")
        return True
    else:
        print("Sorry, not enough stock available..")
        return False


def product_choice(section, products):
    prod_choice = int(input("\nWhich would you like to buy? (type 1 or 2): ")) - 1
    qty = int(input("\nHow many would you like to buy? "))

    if 0 <= prod_choice < len(products):
        handle_purchase(products[prod_choice], qty)
    else:
        print("Invalid Choice, please enter a valid number..")

File: c++/Product.cpp
#include "Product.h"

// Product class implementation
Product::Product(string n, double p, int s) : name(n), price(p), stock(s) {}
// Destructor
Product::~Product() {}

//setters and getters for base product details
string Product::getName() const { return name; }
void Product::setName(const string& n) { name = n; }
double Product::getPrice() const { return price; }
void Product::setPrice(double p) { price = p; }
int Product::getStock() const { return stock; }
void Product::setStock(int s) { stock = s; }

//function to update inventory
void Product::updateStock(int quantity) {
    if (quantity > stock) {
        cout << "Not enough in stock..\n";
    } else {
        stock -= quantity;
    }
}

//function to display product details, can be overridden
void Product::display() const {
    cout << "Product: " << name << "\nPrice: $" << price << "\nStock: " << stock << "\n";
}

// Derived classes implementation
Volleyball::Volleyball(string n, double p, int s, string m, int sz)
    : Product(n, p, s), material(m), size(sz) {}

//override display method
void Volleyball::display() const {
    Product::display();
    cout << "Material: " << material << "\nSize: " << size << " inches\n\n";
}

volleyballNet::volleyballNet(string n, double p, int s, string t)
    : Product(n, p, s), type(t) {}

void volleyballNet::display() const {
    Product::display();
    cout << "Type: " << type << "\n\n";
}

Paddle::Paddle(string n, double p, int s, string st)
    : Product(n, p, s), style(st) {}

void Paddle::display() const {
    Product::display();
    cout << "Style: " << style << "\n\n";
}

Table::Table(string n, double p, int s, string m)
    : Product(n, p, s), material(m) {}

void Table::display() const {
    Product::display();
    cout << "Material: " << material << "\n\n";
}

Ball::Ball(string n, double p, int s)
    : Product(n, p, s) {}

Tent::Tent(string n, double p, int s, string sz, string d, string wr)
    : Product(n, p, s), size(sz), durability(d), weatherRating(wr) {}

void Tent::display() const {
    Product::display();
    cout << "Size: " << size << "\nDurability: " << durability << "\nWeather Rating: " << weatherRating << "\n\n";
}

SleepingBag::SleepingBag(string n, double p, int s, string sz)
    : Product(n, p, s), size(sz) {}

void SleepingBag::display() const {
    Product::display();
    cout << "Size: " << size << "\n\n";
}

SoccerBall::SoccerBall(string n, double p, int s, string m)
    : Product(n, p, s), material(m) {}

void SoccerBall::display() const {
    Product::display();
    cout << "Material: " << material << "\n\n";
}

GoalNet::GoalNet(string n, double p, int s, string sz)
    : Product(n, p, s), size(sz) {}

void GoalNet::display() const {
    Product::display();
    cout << "Size: " << size << "\n\n";
}

//displays product details by extracting info from pointer to product
void displayProductDetails(Product* product) {
    if (product == nullptr) {
        cout << "Invalid Product.\n";
        return;
    }
    product->display();
}

//function to handle purchasing from user
bool handlePurchase(Product* product, int quantity) {
    //check if product is in stock
    if (product->getStock() >= quantity) {
        //calculate total price
        double totalPrice = product->getPrice() * quantity;
        cout << "\nTotal Price: $" << totalPrice << "\n";
        //update stock inventory
        product->updateStock(quantity);
        cout << "Updated Stock: " << product->getStock() << " remaining\n";
        return true;
    } else {
        cout << "Sorry, not enough stock available..\n";
        return false;
    }
}

void productChoice(const string& section, vector<Product*>& products) {
    int prodChoice;
    int qty;

    //ask user which product they want
    cout << "\nWhich  would you like to buy? (type 1 or 2): ";
    cin >> prodChoice;

    //ask user how many they want
    cout << "\nHow many would you like to buy? ";
    cin >> qty;

    //handle user input
    if (prodChoice >= 1 && prodChoice <= products.size()) {
        //vector count starts at 0 so whichever product they choose, subtract one
        //to get correct product from vector
        handlePurchase(products[prodChoice - 1], qty);
    }
    else {
        cout << "Invalid Choice, please enter a valid number..\n";
    }
}